File: src/Entity/RendezVous.php
<?php

namespace App\Entity;

use App\Enum\ModeType;
use App\Enum\MotifType;
use App\Enum\StatutType;
use App\Repository\RendezVousRepository;
use Doctrine\DBAL\Types\Types;
use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Validator\Constraints as Assert;
use Symfony\Component\Validator\Context\ExecutionContextInterface;


#[ORM\Entity(repositoryClass: RendezVousRepository::class)]
class RendezVous
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    private ?int $id = null;

    #[ORM\Column(type: Types::DATETIME_MUTABLE)]
    #[Assert\NotNull(message: "La date et l'heure du rendez-vous sont obligatoires.")]
    #[Assert\GreaterThan("today", message: "La date du rendez-vous doit être dans le futur.")]
    private ?\DateTimeInterface $date_heure = null;

    #[ORM\Column(length: 255)]
    private ?MotifType $motif = null;

    #[ORM\Column(length: 255)]
    private ?StatutType $statut = null;

    #[ORM\Column(length: 255)]
    private ?ModeType $mode = null;

    #[ORM\Column(type: "string", length: 255,nullable: false)]
    #[Assert\NotBlank(message: "Le commentaire ne peut pas être vide.")]
    #[Assert\Length(max: 255, maxMessage: "Le commentaire ne doit pas dépasser 255 caractères.")]
    private ?string $commantaire = null;

    #[ORM\ManyToOne(inversedBy: 'id_medecin')]
    private ?Utilisateur $id_patient = null;

    #[ORM\ManyToOne(inversedBy: 'rendezVouses')]
    private ?Utilisateur $id_medecin = null;

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getDateHeure(): ?\DateTimeInterface
    {
        return $this->date_heure;
    }

    public function setDateHeure(\DateTimeInterface $date_heure): static
    {
        $this->date_heure = $date_heure;

        return $this;
    }

    public function getMotif(): ?MotifType
    {
        return $this->motif;
    }

    public function setMotif(MotifType $motif): static
    {
        $this->motif = $motif;

        return $this;
    }

    public function getStatut(): ?StatutType
    {
        return $this->statut;
    }

    public function setStatut(StatutType $statut): static
    {
        $this->statut =  $statut;

        return $this;
    }

    public function getMode(): ?ModeType
    {
        return $this->mode;
    }

    public function setMode(ModeType $mode): static
    {
        $this->mode = $mode;

        return $this;
    }

    public function getCommantaire(): ?string
    {
        return $this->commantaire;
    }

    public function setCommantaire(string $commantaire): self
    {
        $this->commantaire = $commantaire;

        return $this;
    }

    public function getIdPatient(): ?Utilisateur
    {
        return $this->id_patient;
    }

    public function setIdPatient(?Utilisateur $id_patient): static
    {
        $this->id_patient = $id_patient;

        return $this;
    }

    public function getIdMedecin(): ?Utilisateur
    {
        return $this->id_medecin;
    }

    public function setIdMedecin(?Utilisateur $id_medecin): static
    {
        $this->id_medecin = $id_medecin;

        return $this;
    }

    #[Assert\Callback('validateDateHeure')]
    public function validateDateHeure(ExecutionContextInterface $context ): void
{
    if ($this->date_heure === null) {
        return;
    }

    $now = new \DateTime();
    $hour = (int) $this->date_heure->format('H');

    if ($this->date_heure < $now) {
        $context->buildViolation("La date du rendez-vous doit être dans le futur.")
            ->atPath('date_heure')
            ->addViolation();
    }

    if ($hour < 8 || $hour > 16) {
        $context->buildViolation("L'heure du rendez-vous doit être entre 08:00 et 16:00.")
            ->atPath('date_heure')
            ->addViolation();
    }
}

public function __construct()
{
    $this->date_heure = new \DateTime(); // Définit la date actuelle par défaut
}



}

File: src/main/java/org/example/entities/RendezVous.java
package org.example.entities;

import org.example.enums.Mode;
import org.example.enums.Motif;
import org.example.enums.Statut;

import java.time.LocalDateTime;

public class RendezVous {
    private int id;

    private LocalDateTime dateHeure;

    private Motif motif;

    private Statut statut;

    private Mode mode;

    private String commentaire;

    private Utilisateur medecin;

    private Utilisateur patient;

    public RendezVous(int id) {
        this.id = id;
    }

    public RendezVous(LocalDateTime dateHeure, Motif motif, Statut statut, Mode mode, String commentaire, Utilisateur medecin, Utilisateur patient) {
        this.dateHeure = dateHeure;
        this.motif = motif;
        this.statut = statut;
        this.mode = mode;
        this.commentaire = commentaire;
        this.medecin = medecin;
        this.patient = patient;
    }

    public RendezVous(int id, LocalDateTime dateHeure, Motif motif, Statut statut, Mode mode, String commentaire, Utilisateur medecin, Utilisateur patient) {
        this.id = id;
        this.dateHeure = dateHeure;
        this.motif = motif;
        this.statut = statut;
        this.mode = mode;
        this.commentaire = commentaire;
        this.medecin = medecin;
        this.patient = patient;
    }

    public RendezVous(LocalDateTime dateHeure) {
        this.dateHeure = dateHeure;
    }
    public RendezVous() {

    }


    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public LocalDateTime getDateHeure() {
        return dateHeure;
    }

    public void setDateHeure(LocalDateTime dateHeure) {

        this.dateHeure = dateHeure;
    }

    public Motif getMotif() {
        return motif;
    }

    public void setMotif(Motif motif) {

        this.motif = motif;
    }

    public Statut getStatut() {
        return statut;
    }

    public void setStatut(Statut statut) {

        this.statut = statut;
    }

    public Mode getMode() {
        return mode;
    }

    public void setMode(Mode mode) {

        this.mode = mode;
    }

    public String getCommentaire() {
        return commentaire;
    }

    public void setCommentaire(String commentaire) {


        this.commentaire = commentaire;
    }

    public Utilisateur getMedecin() {
        return medecin;
    }

    public void setMedecin(Utilisateur medecin) {

        this.medecin = medecin;
    }

    public Utilisateur getPatient() {
        return patient;
    }

    public void setPatient(Utilisateur patient) {
        this.patient = patient;
    }

    @Override
    public String toString() {
        return "RendezVous{" +
                "id=" + id +
                ", dateHeure=" + dateHeure +
                ", motif='" + motif + '\'' +
                ", statut='" + statut + '\'' +
                ", mode='" + mode + '\'' +
                ", commentaire='" + commentaire + '\'' +
                ", medecin=" + medecin +
                ", patient=" + patient +
                '}';
    }
}
